src/amo/utils/addonDownload.ts

```diff
@@ -4,14 +4,61 @@ import * as vscode from "vscode";
 
 import constants from "../../config/config";
 
-export async function downloadAddon(fileId: string, path: string) {
+async function fetchDownloadFile(fileId: string) {
   const url = `${constants.downloadBaseURL}${fileId}`;
   const response = await fetch(url);
-
-  if (response.ok) {
-    const dest = fs.createWriteStream(path, { flags: "w" });
-    dest.write(await response.buffer());
-    dest.close();
-    vscode.window.showInformationMessage("Download complete");
+  if (!response.ok) {
+    await vscode.window
+      .showErrorMessage(
+        `(Status ${response.status}): Could not fetch download file.`,
+        { modal: true },
+        { title: "Try Again" },
+        { title: "Fetch New Addon" }
+      )
+      .then(async (action) => {
+        if (action?.title === "Try Again") {
+          return await fetchDownloadFile(fileId);
+        } else if (action?.title === "Fetch New Addon") {
+          vscode.commands.executeCommand("assay.get");
+          throw new Error("Process restarted");
+        } else {
+          throw new Error("Request failed");
+        }
+      });
   }
+  return response;
+}
+
+export async function downloadAddon(fileId: string, path: string) {
```

Comment:
the code in this appears very similar to `fetchDownloadFile` - is there scope for abstracting some of it out into a common function?  (DRY principle)

Reply:
yes, a single error function seems a good move.  The following chunk of code appears identical in at least `fetchDownloadFile` and `downloadAddon` - although it's written with `await` in once place and `.then` in the other.

```javascript
        const action = await vscode.window.showErrorMessage(
          `Could not download addon to ${path}.`,
          { modal: true },
          { title: "Try Again" },
          { title: "Fetch New Addon" }
        );

        if (action?.title === "Try Again") {
          return await downloadAddon(fileId, path);
        } else if (action?.title === "Fetch New Addon") {
          vscode.commands.executeCommand("assay.get");
          throw new Error("Process restarted");
        } else {
          throw new Error("Download failed");
        }
```

Can you look at doing it for this patch?  It's always better to smaller amounts of code in the first place, unless there is a time factor.